### ISSUE TITLE
Allow loading settings from JSON?

### ISSUE THREAD

**Issue by RReverser** [emscripten-core/emscripten#14317]
Saw this on twitter: https://twitter.com/hardfist_1/status/1397332063330521090

> really wish emscripten supports json configure like tsconfig.json，writing all configuration in a script is awful

I think it's a really good idea to allow specifying at least common settings in a JSON file instead of a long list of shell parameters, and would make configuration easier.

cc @sbc100 as you've been working on loading specific settings from newline-delimited files, and this seems like a higher-level extension of the same idea

**Comment by sbc100** [comment]
Interesting.  So you would do something like `emcc foo.c --settings=foo.json`?  

That seems fairly simple to implement... but I do worry about adding more features like this.   At what level do imagine a user would use this?   Presumably it would still need to be per-module?  (i.e. not `$HOME/.emsettings` or even `project/.emsettings`).

Do you know about the existence of the response file processing that already exists today?  Perhaps that solves this problem without adding new tech?  

```
$ emcc foo.c @mysettings.txt
$ cat mysettings.txt
-sFOO
-O3
-g
```

That file format also seems less noisy and easier to read that trying to convert each setting into json... it also already exists and is something that we shared with existing other native tools.

**Comment by sbc100** [comment]
This response file works at a higher level so you could still have `-sEXPORTED_FUNCTIONS=@exports.txt` within your `mysettings.txt` file.

**Comment by RReverser** [comment]
Interesting, I didn't know you could pass top-level settings like that.

I think JSON is interesting in many regards, particularly it might be more familiar to Web devs, but having even this kind of file-based configuration is already useful.

Is it documented somewhere? I don't see it on https://emscripten.org/docs/tools_reference/emcc.html.

**Comment by RReverser** [comment]
> I think JSON is interesting in many regards, particularly it might be more familiar to Web devs, but having even this kind of file-based configuration is already useful.

Another thing is, if I understand correctly, for Emscripten-specific settings one would always need to join `-sFOO` or put `-s` and `FOO` on different lines? If so, that certainly feels less readable than

```json
{
  "FOO": 1,
  "BAR": 2
}
```

**Comment by sbc100** [comment]
No, I think newline acts like space so all of these should be the same:

```
-s
FOO=1
```

```
-sFOO=1
```

```
-s FOO=1`
```

**Comment by sbc100** [comment]
The downside of limiting this to just settings is of course would be that you couldn't use such a mechanism for any of the non setting arguments.

Personally I see no need to introduce json into these things.   Even in one-line-per-setting file format wouldn't it be simpler just to do:

```
FOO=1
BAR=2
```

or  if we want to match settings.js itself (which historically I believe was something that one would edit globally to perform these kinds of global settings mods):

```
var FOO = 1;
var BAR = 2;
```

(only joking about that last option!).

Personally I'd rather just stick to the existing response files rather than adding yet more ways to do things.

**Comment by sbc100** [comment]
The `@response_file` thing works for many/all of our tools (such as `emcc` and `emar`) and also can be used directly with gcc and clang and other llvm tools.   I agree we should document it, or at least reference it somehow, in the emcc docs.

**Comment by donalffons** [comment]
Just as a note: I'm using [YAML files](https://github.com/donalffons/opencascade.js/blob/master/src/builds/opencascade.full.yml#L160) in my project to do this :slightly_smiling_face:. If there would be an "official" format for this, that could be neat...

**Comment by sbc100** [comment]
@donalffons, is there some reason you couldn't switch to just using the raw command line in a @response file?    That format is already supported.     Less quoting, less typing, easier coping and pasting too and from the actual command line?

**Comment by sbc100** [comment]
> @donalffons, is there some reason you couldn't switch to just using the raw command line in a @response file? That format is already supported. Less quoting, less typing, easier coping and pasting too and from the actual command line?

Oh I see you have emcc options as just part of a bigger yaml build system.   Seems like your approach is fine.  Seems like even if we decided on json you wouldn't want to shoehorn json into your existing yaml file.   

Same with bazel build file.. these existing systems know about command line args already and probably don't want a new format.

**Comment by RReverser** [comment]
> Even in one-line-per-setting file format wouldn't it be simpler just to do:

Does that work today?

**Comment by sbc100** [comment]
> > Even in one-line-per-setting file format wouldn't it be simpler just to do:
> 
> Does that work today?

No.. I was just pointing out that if we invent a new format there are simpler options than json out there.  But as I say, my inclination is to not do anything :)

**Comment by RReverser** [comment]
I think JSON is nice mainly for compatibility with other systems, autocompletion in IDE etc., even if it's more verbose.

E.g. VSCode supports JSON schemes, which automatically links them to corresponding docs and autocompleted names, which is very valuable when working with `tsconfig.json`, `eslint.json`, or any config that can potentially have a hundred of possible settings - Emscripten feels very much in that category.

That is, if IDE provided a way to discover those settings, maybe website versions like Surma's wouldn't have to be built in the first place :)

**Comment by RReverser** [comment]
Unrelated - this question still stands:

> Is it documented somewhere? I don't see it on emscripten.org/docs/tools_reference/emcc.html.

**Comment by sbc100** [comment]
I think I answered that in https://github.com/emscripten-core/emscripten/issues/14317#issuecomment-849878274

**Comment by RReverser** [comment]
Ah missed that comment, sorry. Yeah, as someone not _that_ familiar with those tool, it was certainly news to me that it's already supported across the board.
